httpclient/client.go

```diff
@@ -0,0 +1,71 @@
+package httpclient
+
+import (
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/gojekfarm/albatross-client-go/config"
+	"github.com/gojekfarm/albatross-client-go/logger"
+)
+
+type client interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// Client acts as a wrapper around the net/http.Client to take care of exponential retries.
+// TODO: Discuss if we should use go-retryablehttp
+type Client struct {
+	client client
+	retry  *config.Retry
+	logger logger.Logger
+}
+
+// Request executes a given request with the provided retry policy
+// TODO: Handle timeouts and implement retries
+// The send implementation follows the same behaviour as the default implementation
+// The error is returned only in case of network errors, non 2xx codes do not cause errors
+// The assumption is that since it's a json api, all responses should result in a valid
+// json body unless under exceptional circumstances. The users can check the response status code
+// and parse the bytestream accordingly.
+// The client can be extended to handle authentication failures
+func (c *Client) Send(url string, method string, body io.Reader) (*http.Response, []byte, error) {
+	request, err := http.NewRequest(method, url, body)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resp, err := c.client.Do(request)
+	if err != nil {
+		c.logger.Errorf("Error connecting to the albatross api: %s", err)
+		return nil, nil, err
+	}
+
+	defer resp.Body.Close()
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return resp, nil, err
+	}
+
+	// Log all 5xx errors, and returns the resp for additional parsing
+	if resp.StatusCode >= 500 {
+		// With a 5xx, the response most likely won't be parsable
+		c.logger.Errorf("server error for albatross api: %s - %s", url, data)
+		return resp, nil, nil
```

Comment:
This method returns the http response object so that the caller can take appropriate action based on the status code. The error here is reserved for the cases where the client was unable to connect to the API, in other words, a nil error represents that there was a response.